src/common/options.cc

```diff
@@ -7454,6 +7454,10 @@ static std::vector<Option> get_rbd_options() {
     .set_default(0)
     .set_description("time-delay in seconds for rbd-mirror asynchronous replication"),
 
+    Option("rbd_mirroring_max_snapshots", Option::TYPE_UINT, Option::LEVEL_ADVANCED)
+    .set_default(2)
```

Comment:
I wonder if it would make sense to set this as a minimum of 3. Then if you create a new mirroring snapshot and your peer is behind, you could drop the peer from the oldest mirrored snapshot and create it against the new one. That would ensure that the peer (assuming it ever catches up) would end up w/ the most up-to-date copy. Then I think we could possibly drop the need for the force create flag.

Reply:
I also thought that 2 might be too low. As for the idea to automatically unlink the peer from older snapshots (and thus to remove the snapshot if no peers remain), I am not sure how to handle a potential situation when we are removing a snapshot that is being used by rbd-mirror right now for copying data.

Reply:
The rbd-mirror daemon would only ever need two snapshots to apply the diff. Once we have three snapshots, we know the rbd-mirror daemon isn't currently using the third one. There is a potential for a small race which we could protect against in the unlink peer case if we provided an optional assertion that snapshot XYZ is still attached to the peer (since rbd-mirror would unlink from the oldest snapshot prior to applying the delta between the next two).

Reply:
Ok, it seems to work. Thanks!

Reply:
@dillaman I suppose that when the rbd-mirror sees three snapshots (sync points) A, B, C, it will delete B and start copy C-A diff. Then when a user creates a new snapshot and hits the limit and wants to delete (unlink) a snapshot it should always choose the middle one.This way we may race only for snapshot deletion.

Alternatively, the rbd-mirror could always copy the diff for adjacent snpashots, i.e. copy B-A diff, delete A, copy C-B diff, delete B. Then the user when creating a new snapshot and hitting the limit should always unlink the newest snapshot.

I prefer the first variant as it looks like more optimal to copy the diff between the oldest and the newest snapshots. Also in the second variant it seems there is still a race when the  user sees 3 snapshots (A, B, C) and is going to delete the last one (C), and at that time the rbd-mirror finishes copying B-A diff, deletes A and starts copying C-B.

Then if we are going to use the first variant, i.e. deleting the middle snapshot, the `predecessor_snap_id` seems not useful any more, and I'd rather not add it to the snapshot namespace?

Reply:
Yeah, we could do a C-A diff. I think the one downside is that since we will need to rollback on forced promotion after partial replay, you will potentially go further back in time -- but that would really only occur on essentially random IO workloads where it's mostly different blocks being updated (which probably doesn't match real-world workloads since disk schedulers try to avoid seeks).

Agree that the `predecessor_snap_id` doesn't matter -- but we will still need a way to track potential split-brain.